apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts

```diff
@@ -0,0 +1,346 @@
+    }
+
+    if (newGoverningAstDeclaration) {
+      newGoverningAstDeclaration.astSymbol.notifyAnalyzed();
+    }
+  }
+
+  // tslint:disable-next-line:no-unused-variable
+  private _fetchAstDeclaration(node: ts.Node): AstDeclaration | undefined {
+    const astSymbol: AstSymbol | undefined = this._fetchAstSymbolForNode(node);
+    if (!astSymbol) {
+      return undefined;
+    }
+
+    const astDeclaration: AstDeclaration | undefined
+      = this._astDeclarationsByDeclaration.get(node);
+    if (!astDeclaration) {
+      throw new Error('Program Bug: Unable to find constructed AstDeclaration');
+    }
+
+    return astDeclaration;
+  }
+
+  private _fetchAstSymbolForNode(node: ts.Node): AstSymbol | undefined {
+    if (!this.isAstDeclaration(node)) {
+      return undefined;
+    }
+
+    const symbol: ts.Symbol | undefined = this.tryGetSymbolForNode(node);
+    if (!symbol) {
+      throw new Error('Program Bug: Unable to find symbol for node');
+    }
+
+    return this._fetchAstSymbol(symbol, true);
+  }
+
+  private _fetchAstSymbol(symbol: ts.Symbol, addIfMissing: boolean): AstSymbol | undefined {
+    const followAliasesResult: IFollowAliasesResult = SymbolAnalyzer.followAliases(symbol, this._typeChecker);
+
+    const followedSymbol: ts.Symbol = followAliasesResult.followedSymbol;
+    if (followedSymbol.flags & (ts.SymbolFlags.TypeParameter | ts.SymbolFlags.TypeLiteral | ts.SymbolFlags.Transient)) {
+      return undefined;
+    }
+
+    if (followAliasesResult.isAmbient) {
+      return undefined;
+    }
+
+    let astSymbol: AstSymbol | undefined = this._astSymbolsBySymbol.get(followedSymbol);
+
+    if (!astSymbol) {
+      if (!followedSymbol.declarations || followedSymbol.declarations.length < 1) {
+        throw new Error('Program Bug: Followed a symbol with no declarations');
+      }
+
+      for (const declaration of followedSymbol.declarations || []) {
+        if (!this.isAstDeclaration(declaration)) {
+          throw new Error(`Program Bug: The "${followedSymbol.name}" symbol uses the construct`
+            + ` "${ts.SyntaxKind[declaration.kind]}" which may be an unimplemented language feature`);
+        }
+      }
+
+      const astImport: AstImport | undefined = followAliasesResult.astImport;
+
+      if (astImport) {
+        astSymbol = this._entriesByImportKey.get(astImport.key);
+
+        if (astSymbol) {
+          // We didn't find the entry using followedSymbol, but we did using importPackageKey,
+          // so add a mapping for followedSymbol; we'll need it later when renaming identifiers
+          this._astSymbolsBySymbol.set(followedSymbol, astSymbol);
+        }
+      }
+
+      if (!astSymbol) {
+        // None of the above lookups worked, so create a new entry
+        astSymbol = new AstSymbol({
+          localName: followAliasesResult.localName,
+          followedSymbol: followAliasesResult.followedSymbol,
```

Comment:
>followedSymbol [](start = 10, length = 14)

originalSymbol? "Followed" is a bit weird

Reply:
I'll handle this in a separate PR. I need to think of a better word for "following all the paths to their unique endpoint".  "Dealias" is apparently a legitimate mathematical word, but pretty weird to type.  "Normalize" is sort of the right idea, but I need an unambiguous jargon for this important part of the algorithm.

@nickpape-msft @iclanton

---
In reply to: [174001612](https://github.com/Microsoft/web-build-tools/pull/560#discussion_r174001612) [](ancestors = 174001612)